### PR TITLE
 Add shard state into genesis and fix shard state hashing bug

### PR DESCRIPTION
**Pull request by rlan35**

### CODE REVIEW

File: core/genesis.go [hunk 1]
```diff
@@ -251,7 +257,7 @@ func (g *Genesis) ToBlock(db ethdb.Database) *types.Block {
 		Coinbase:       g.Coinbase,
 		Root:           root,
 		ShardStateHash: g.ShardStateHash,
-		// ShardState is absent in epoch block; genesis shard state is implied
+		ShardState:     shardStateBytes,
```
Comment (chaosma) [inline core/genesis.go]: Is this only in genesis block or also in other blocks?

File: core/genesis.go [hunk 2]
```diff
@@ -303,12 +315,11 @@ func GenesisBlockForTesting(db ethdb.Database, addr common.Address, balance *big
...
 	return &Genesis{
-		Config:     params.MainnetChainConfig,
-		Nonce:      66,
-		ExtraData:  hexutil.MustDecode("0x11bbe8db4e347b4e8c937c1c8370e4b5ed33adb3db69cbdb7a38e1e50b1b82fa"),
-		GasLimit:   5000,
-		Difficulty: big.NewInt(17179869184),
-		Alloc:      decodePrealloc("empty"),
+		Config:    params.MainnetChainConfig,
+		Nonce:     66,
+		ExtraData: hexutil.MustDecode("0x11bbe8db4e347b4e8c937c1c8370e4b5ed33adb3db69cbdb7a38e1e50b1b82fa"),
```
Comment (chaosma) [inline core/genesis.go]: what is the extra data here ?

File: core/types/block.go [hunk 3]
```diff
@@ -95,8 +95,8 @@ type Header struct {
 	Vdf              [258]byte   `json:"vdf"`
 	VdfProof         [258]byte   `json:"vdfProof"`
 	ShardStateHash   common.Hash `json:"shardStateRoot"`
-	ShardState       ShardState  `json:"shardState"`
-	CrossLinks       [][]byte    `json:"crossLinks"`
+	ShardState       []byte      `json:"shardState"`
```
Comment (chaosma) [inline core/types/block.go]: will all blocks save shard state or just epoch blocks or just genesis block

Comment (chaosma) [reply core/types/block.go]: what is crosslink format, change from [][]byte to []byte?

**Comment by LeoHChen** [comment]
Please don't use long running branch. Open a new branch for every PR. https://github.com/harmony-one/harmony-open/tree/master/git-cheat-sheet

**Pull request merged**